checkoff

```diff
@@ -1,167 +1,162 @@
+        'student belongs to track {} but lab is for track {}'.format(student_track, lab_track))
 
-    if user not in roster:
-        print('{} not in class roster'.format(user))
-        return 1
+    with db() as c:
+        c.execute('INSERT INTO `checkoffs` (`lab`, `student`, `facilitator`)'
+                  'VALUES (%s, %s, %s)', (lab, user, facilitator))
 
-    completed_labs = []
-    if user in users:
-        completed_labs = sorted(users[user])
+    print('Checked off {} for lab {}'.format(user, lab))
+    _mail_checkoff(user, lab, data)
 
-    print('{} checked off for labs: {}'.format(user, completed_labs))
 
-
-def list_labs(args):
+def lab_completed_students(args, data):
     lab = args.lab
-    _check_lab_id(lab)
 
-    data = _load_data()
-    users = data['users']
+    with db() as c:
+        c.execute('SELECT `student`, `facilitator`, `timestamp` FROM `checkoffs`'
+                  'WHERE `lab` = %s', lab)
 
-    for student in users:
-        if lab in users[student]:
-            print(student)
+    print('Checkoffs for lab {} ({})'.format(lab, data['labs'][lab]['fullname']))
 
+    for i in c:
+        print('{} checked off by {} on {}'.format(
+            i['student'], i['facilitator'], i['timestamp']
+        ))
 
-def _acquire_lock():
-    # Atomically check and create lockfile
-    # https://stackoverflow.com/a/33223732/5194666
-    try:
-        os.open(LOCK_FILE_PATH, os.O_CREAT | os.O_EXCL)
-        return True
-    except FileExistsError:
-        return False
 
+def student_completed_labs(args, data):
+    user = args.user
+
+    with db() as c:
+        c.execute('SELECT `lab`, `facilitator`, `timestamp` FROM `checkoffs`'
+                  'WHERE `student` = %s', user)
 
-def _release_lock():
-    os.remove(LOCK_FILE_PATH)
+    print('Checkoffs for student {} (track: {})'.format(user, data['roster'][user]))
 
+    for i in c:
+        print('Lab {} checked off by {} on {}'.format(
+            i['lab'], i['facilitator'], i['timestamp']
+        ))
 
-def _mail_checkoff(user, lab):
+
+def _mail_checkoff(user, lab, data):
     # subprocess needs input in bytes
-    msg = 'Congratulations {}, you have been checked off for lab {}'.format(user, lab).encode()
-    subject = 'Unix Sysadmin Decal Checkoff for Lab {}'.format(lab)
+    msg = 'Congratulations {}, you have been checked off for lab {} ({})'.format(
+        user, lab, data['labs'][lab]['fullname'],
+    ).encode()
+    subject = 'Unix SysAdmin DeCal Checkoff for Lab {}'.format(lab)
     recipient = '{}@ocf.berkeley.edu'.format(user)
     cc = 'decal+checkoffs@ocf.berkeley.edu'
-    subprocess.run(['mail', '-s', subject, '-c', cc, recipient], stdout=subprocess.PIPE, input=msg)
-
-
-def _load_data():
-    with open(DATA_PATH, 'r') as f:
-        data = json.load(f)
-    return data
-
-
-def _write_data(data):
-    with open(DATA_PATH, 'w') as f:
-        json.dump(data, f)
+    subprocess.run(['mail', '-s', subject, '-c', cc, recipient],
```

Comment:
Something from [ocflib](https://github.com/ocf/ocflib/blob/61e8b085e3c877eed42d844cc16228a29fcc917b/ocflib/misc/mail.py#L63) would've been nice for this, but there is no cc option =/

Reply:
if you want to add a cc option to ocflib, I'll update this script to match

Reply:
[done](https://github.com/ocf/ocflib/pull/114).